corehq/apps/commtrack/models.py

```diff
@@ -551,38 +547,39 @@ def get_data(self, value):
         return values
 
 
-def sync_location_supply_point(loc):
-    """
-    This method syncs the location/supply point connection
-    and is triggered whenever a location is edited or created.
-    """
-    # circular import
-    from corehq.apps.domain.models import Domain
+def _make_location_admininstrative(location):
+    supply_point_id = location.supply_point_id
+    if supply_point_id:
+        close_case(supply_point_id, location.domain, const.COMMTRACK_USERNAME)
+    location.supply_point_id = None  # this will be saved soon anyways
 
-    domain = Domain.get_by_name(loc.domain)
-    if not domain.commtrack_enabled:
-        return
 
-    def _needs_supply_point(loc, domain):
-        """Exclude administrative-only locs"""
-        return loc.location_type in [loc_type.name for loc_type in domain.location_types if not loc_type.administrative]
+def _reopen_or_create_supply_point(location):
+    from .dbaccessors import get_supply_point_by_location_id
+    supply_point = get_supply_point_by_location_id(location.domain, location.location_id)
+    if supply_point:
+        if supply_point and supply_point.closed:
+            for action in supply_point.actions:
+                if action.action_type == 'close':
+                    action.xform.archive(user_id=const.COMMTRACK_USERNAME)
+        supply_point.update_from_location(location)
+        return supply_point
+    else:
+        return SupplyInterface.create_from_location(location.domain, location)
 
-    if _needs_supply_point(loc, domain):
-        supply_point = get_supply_point_case_by_location(loc)
-        if supply_point:
-            supply_point.update_from_location(loc)
-            updated_supply_point = supply_point
-        else:
-            updated_supply_point = SupplyInterface.create_from_location(loc.domain, loc)
 
-        # need to sync this sp change to the sql location
-        # but saving the doc will trigger a loop
-        try:
-            sql_loc = SQLLocation.objects.get(location_id=loc.location_id)
-            sql_loc.supply_point_id = updated_supply_point._id
-            sql_loc.save()
-        except SQLLocation.DoesNotExist:
-            pass
+def sync_supply_point(location):
+    # Called on location.save()
+    domain = Domain.get_by_name(location.domain)
+    if not domain.commtrack_enabled:
+        return None
+
+    if location.location_type.administrative:
+        _make_location_admininstrative(location)
+        return None
+    else:
+        updated_supply_point = _reopen_or_create_supply_point(location)
+        return updated_supply_point._id
```

Comment:
This function is where the main logic change is - now it handles closing and reopening supply-point cases.